File: 2019/day01.py
def load():
    with open("day01.txt") as f:
        xs = f.readlines()
        return [int(x) for x in xs]


def part1(xs):
    return sum([x // 3 - 2 for x in xs])


def part2(xs):

    def go(x, acc):
        if x == 0:
            return acc
        else:
            y = x // 3 - 2
            z = y if y > 0 else 0
            acc.append(z)
            return go(z, acc)

    return sum([sum(go(x, [])) for x in xs])


class TestFoo:

    def test_part1(self):
        assert part1(load()) == 3087896

    def test_part2(self):
        assert part2(load()) == 4628989

File: 2021/elixiraoc/test/day01.exs
defmodule Day01 do
  def data do
    File.stream!("./test/day01.txt")
    |> Stream.map(&String.trim/1)
    |> Stream.map(&String.to_integer/1)
  end

  def part1(xs, n \\ 1) do
    xs
    |> Stream.chunk_every(n, 1, :discard)
    |> Stream.chunk_every(2, 1, :discard)
    |> Enum.count(fn [x, y] -> Enum.sum(y) > Enum.sum(x) end)
  end

  def part2(xs) do
    part1(xs, 3)
  end
end

defmodule Day01Test do
  use ExUnit.Case
  doctest Elixiraoc

  test "part1" do
    assert Day01.data() |> Day01.part1() == 1482
  end

  test "part2" do
    assert Day01.data() |> Day01.part2() == 1518
  end
end
